stdlib/public/core/StringCreate.swift

```diff
@@ -298,4 +298,53 @@ extension String {
       String._uncheckedFromUTF8($0)
     }
   }
+
+  @usableFromInline
+  @available(SwiftStdlib 5.10, *)
+  internal static func _validate<Encoding: Unicode.Encoding>(
+    _ input: UnsafeBufferPointer<Encoding.CodeUnit>,
+    as encoding: Encoding.Type
+  ) -> String? {
+  fast: // fast-path
+    if encoding.CodeUnit.self == UInt8.self {
+      let bytes = _identityCast(input, to: UnsafeBufferPointer<UInt8>.self)
+      let isASCII: Bool
+      if encoding.self == UTF8.self {
+        guard case .success(let info) = validateUTF8(bytes) else { return nil }
+        isASCII = info.isASCII
+      } else if encoding.self == Unicode.ASCII.self {
+        guard _allASCII(bytes) else { return nil }
+        isASCII = true
+      } else {
+        break fast
+      }
+      return String._uncheckedFromUTF8(bytes, asciiPreScanResult: isASCII)
+    }
```

Comment:
This is the only part that benefits from contiguous storage, because `validateUTF8` and `_allASCII` require an unsafe buffer pointer. The second part of the function (the slow path, which transcodes) doesn't need contiguous storage at all, and is basically duplicated code from the initialiser.

I would suggest splitting in to separate functions, and making the call to the contiguous UTF8/ASCII one directly from the initialiser. Since the initialiser is inlinable, I'd expect the compiler could specialise this to a direct call to the UTF8/ASCII fast-path in most cases.

The comment that I made above about performing a dry run and writing directly to string storage would then apply to the second function (the slow path), and we could remove the transcoding in the initialiser.

Reply:
Sorry for the delay, I forgot to respond to this earlier. It is superficially true that the two slow paths are similar at this time, but we have intentionally made the `_validate` function non-inlinable in order to have room to improve it using API that is not yet public. In particular, the current public API doesn't allow for chunked decoding, which would be a significant speedup in a generic context, by significantly amortizing the function call overhead. Getting there from here will be in future work.